man/man1/markdown2pdf.1
.TH MARKDOWN2PDF 1 "January 29, 2011" "Pandoc User Manuals"
.SH NAME
.PP
markdown2pdf - converts markdown-formatted text to PDF, using pdflatex
.SH SYNOPSIS
.PP
markdown2pdf [\f[I]options\f[]] [\f[I]input-file\f[]]...
.SH DESCRIPTION
.PP
\f[C]markdown2pdf\f[] converts \f[I]input-file\f[] (or text from
standard input) from markdown-formatted plain text to PDF, using
\f[C]pandoc\f[] and \f[C]pdflatex\f[].
If no output filename is specified (using the \f[C]-o\f[] option), the
name of the output file is derived from the input file; thus, for
example, if the input file is \f[I]hello.txt\f[], the output file will
be \f[I]hello.pdf\f[].
If the input is read from STDIN and no output filename is specified, the
output file will be named \f[I]stdin.pdf\f[].
If multiple input files are specified, they will be concatenated before
conversion, and the name of the output file will be derived from the
first input file.
.PP
Input is assumed to be in the UTF-8 character encoding.
If your local character encoding is not UTF-8, you should pipe input
through \f[C]iconv\f[]:
.IP
.nf
\f[C]
iconv\ -t\ utf-8\ input.txt\ |\ markdown2pdf
\f[]
.fi
.PP
\f[C]markdown2pdf\f[] assumes that the \f[C]unicode\f[], \f[C]array\f[],
\f[C]fancyvrb\f[], \f[C]graphicx\f[], and \f[C]ulem\f[] packages are in
latex\[aq]s search path.
If these packages are not included in your latex setup, they can be
obtained from \f[C]http://ctan.org\f[].
.SH OPTIONS
.TP
.B -o \f[I]FILE\f[], --output=\f[I]FILE\f[]
Write output to \f[I]FILE\f[].
.RS
.RE
.TP
.B --strict
Use strict markdown syntax, with no extensions or variants.
.RS
.RE
.TP
.B -N, --number-sections
Number section headings in LaTeX output.
(Default is not to number them.)
.RS
.RE
.TP
.B --listings
Use listings package for LaTeX code blocks
.RS
.RE
.TP
.B --template=\f[I]FILE\f[]
Use \f[I]FILE\f[] as a custom template for the generated document.
Implies \f[C]-s\f[].
See the section TEMPLATES in \f[C]pandoc\f[](1) for information about
template syntax.
Use \f[C]pandoc\ -D\ latex\f[] to print the default LaTeX template.
.RS
.RE
.TP
.B -V KEY=VAL, --variable=\f[I]KEY:VAL\f[]
Set the template variable KEY to the value VAL when rendering the
document in standalone mode.
Use this to set the font size when using the default LaTeX template:
\f[C]-V\ fontsize=12pt\f[].
.RS
.RE
.TP
.B -H \f[I]FILE\f[], --include-in-header=\f[I]FILE\f[]
Include (LaTeX) contents of \f[I]FILE\f[] at the end of the header.
Implies \f[C]-s\f[].
.RS
.RE
.TP
.B -B \f[I]FILE\f[], --include-before-body=\f[I]FILE\f[]
Include (LaTeX) contents of \f[I]FILE\f[] at the beginning of the
document body.
.RS
.RE
.TP
.B -A \f[I]FILE\f[], --include-after-body=\f[I]FILE\f[]
Include (LaTeX) contents of \f[I]FILE\f[] at the end of the document
body.
.RS
.RE
.TP
.B --bibliography=\f[I]FILE\f[]
Specify bibliography database to be used in resolving citations.
The database type will be determined from the extension of
\f[I]FILE\f[], which may be \f[C].xml\f[] (MODS format), \f[C].bib\f[]
(BibTeX format), or \f[C].json\f[] (citeproc JSON).
.RS
.RE
.TP
.B --csl=\f[I]FILE\f[]
Specify CSL style to be used in formatting citations and the
bibliography.
If \f[I]FILE\f[] is not found, pandoc will look for it in
.RS
.IP
.nf
\f[C]
$HOME/.csl
\f[]
.fi
.PP
in unix and
.IP
.nf
\f[C]
C:\\Documents\ And\ Settings\\USERNAME\\Application\ Data\\csl
\f[]
.fi
.PP
in Windows.
If the \f[C]--csl\f[] option is not specified, pandoc will use a default
style: either \f[C]default.csl\f[] in the user data directory (see
\f[C]--data-dir\f[]), or, if that is not present, the Chicago
author-date style.
.RE
.TP
.B --data-dir\f[I]=DIRECTORY\f[]
Specify the user data directory to search for pandoc data files.
If this option is not specified, the default user data directory will be
used:
.RS
.IP
.nf
\f[C]
$HOME/.pandoc
\f[]
.fi
.PP
in unix and
.IP
.nf
\f[C]
C:\\Documents\ And\ Settings\\USERNAME\\Application\ Data\\pandoc
\f[]
.fi
.PP
in Windows.
A \f[C]reference.odt\f[], \f[C]epub.css\f[], \f[C]templates\f[]
directory, or \f[C]s5\f[] directory placed in this directory will
override pandoc\[aq]s normal defaults.
.RE
.TP
.B --xetex
Use xelatex instead of pdflatex to create the PDF.
.RS
.RE
.TP
.B --luatex
Use lualatex instead of pdflatex to create the PDF.
.RS
.RE
.SH SEE ALSO
.PP
\f[C]pandoc\f[](1), \f[C]pdflatex\f[](1)
.SH AUTHORS
John MacFarlane, Paulo Tanimoto, and Recai Oktas.


ChangeLog
2006-10-30  Recai Oktaş <roktas@debian.org>

	Ensure that we have always up-to-date version of ChangeLog before creating
	a Debian package.  Make 'ChangeLog' target PHONY to force an update.

2006-10-30  Recai Oktaş <roktas@debian.org>

	Make debian/rules executable to avoid errors in 'distclean' target.

2006-10-30  Recai Oktaş <roktas@debian.org>

	Modifications to create a nice ChangeLog.
	+ Add new file 'gnuify-changelog.pl' grabbed and adapted from subversion
	  distribution.
	+ Add 'ChangeLog' target to create/update ChangeLog.
	+ Update current ChangeLog.

2006-10-30  Recai Oktaş <roktas@debian.org>

	Really fix cabal version detection.

2006-10-30  Recai Oktaş <roktas@debian.org>

	Add pandoc executable to process rules as a dependency.  Make osx_dest,
	osx-pkg distinction more clear.

2006-10-30  Recai Oktaş <roktas@debian.org>

	Parameterise "osx-pkg".

2006-10-30  Recai Oktaş <roktas@debian.org>

	Add a 'osx-pkg' target for the creation of a Mac OSX® package (using
	PackageMaker.App).  This target recursively uses 'install-program'
	target to populate OSX package directory tree.

2006-10-30  Recai Oktaş <roktas@debian.org>

	* cabalize: fix cabal version detection by using a bloody hack.
	* Makefile:
	  + Make 'uninstall-all' depends on 'uninstall-exec'.
	  + Make the default target 'all' depends on $(BINS).  Now, invoking
	    a plain 'make' builds and creates pandoc in the top directory.

2006-10-29  John MacFarlane <jgm@berkeley.edu>

	Bug fixes to make_osx_pkg_dirs.

2006-10-29  John MacFarlane <jgm@berkeley.edu>

	Added a script to make prepare directories for creation of a
	Mac OSX package (using PackageMaker.app).

2006-10-29  John MacFarlane <jgm@berkeley.edu>

	* Extracted installation instructions from README into a separate file,
	INSTALL.  Reason:  Only those who install from source will need the
	information in INSTALL, so it does not need to be included in user
	documentation.  Indeed, it might confuse people who install Pandoc by some
	other method.
	* Minor changes in README (using "Pandoc" as a generic name for the
	system, "'pandoc'" as a name for the executable).

2006-10-29  John MacFarlane <jgm@berkeley.edu>

	Changed date on README.

2006-10-29  John MacFarlane <jgm@berkeley.edu>

	Added a comma in README.

2006-10-29  Recai Oktaş <roktas@debian.org>

	Remove 'install-lib', it causes too much trouble which needs to be dealed
	in ugly ways.
	+ Now, we have 'install-program' and 'install-all' targets.  The former
	  installs only program and user documents, while the latter installs
	  everything: program, user documents, library files and library documents.
	+ Behaviour of 'install' target hasn't changed, it calls 'install-program'.
	+ Update README accordingly.
	+ Implement missing uninstall-* targets (including 'uninstall-all' and
	  'uninstall-program' for the new targets).
	+ Call 'install-program' in debian/rules explicitly to prevent confusions.

2006-10-29  John MacFarlane <jgm@berkeley.edu>

	Changes to RTF writer:
	+ use Helvetica instead of Times New Roman as default font
	+ specify \f0 in every \pard; otherwise font sizes are not registered properly
	+ modify test of RTF writer accordingly

2006-10-29  Recai Oktaş <roktas@debian.org>

	Portability fix.  Remove rmdir(1)'s '--ignore-fail-on-non-empty' option, as
	it is unsupported in BSD version.  Note that the relevant line has already
	started with a '-' which means that make(1) will ignore any error.

2006-10-28  John MacFarlane <jgm@berkeley.edu>

	+ Changed debian/copyright to reflect new website location.

2006-10-28  Recai Oktaş <roktas@debian.org>

	Remove ghc66-branch to reduce clutter.

2006-10-28  Recai Oktaş <roktas@debian.org>

	Merge trunk with ghc66-branch.

2006-10-28  Recai Oktaş <roktas@debian.org>

	Tag current trunk as pandoc-0.21.

2006-10-28  Recai Oktaş <roktas@debian.org>

	Portability fixes:
	+ '+' meta character apparently doesn't work with BSD's sed (without using
	  an extra option).  Just removing this character should be sufficient, as
	  we have now used an unquoted variable at the problematic sed line (see
	  the previous commit) and there is no extra space to remove.
	+ '-D' option of BSD's install(1) doesn't work as expected.  Modify code to
	  not use this option at all.

2006-10-28  Recai Oktaş <roktas@debian.org>

	Unquote a variable in cabalize to have the problematic leading and trailing
	spaces removed in shell expansion.

2006-10-28  John MacFarlane <jgm@berkeley.edu>

	Acknowledge roktas's contributions in README.

2006-10-28  John MacFarlane <jgm@berkeley.edu>

	+ fixed html2markdown usage message, in which w3m occured twice
	+ added usage documentation to beginning of fillTemplates.pl

2006-10-27  Recai Oktaş <roktas@debian.org>

	Merge ghc66-branch with the latest changes in trunk.

2006-10-27  Recai Oktaş <roktas@debian.org>

	Use @HS_SOURCE_DIRS@ in substitution.

2006-10-27  Recai Oktaş <roktas@debian.org>

	Handle 'Hs-Source-Dir[s]*' option in cabalize to get rid of warnings.

2006-10-27  Recai Oktaş <roktas@debian.org>

	Minor corrections.

2006-10-27  Recai Oktaş <roktas@debian.org>

	Add "-o | --output" option to markdown2pdf, update man file.

2006-10-27  John MacFarlane <jgm@berkeley.edu>

	Wording changes in man files.

2006-10-27  Recai Oktaş <roktas@debian.org>

	Merge ghc66-branch with trunk.  Just a future note for this operation:
	+ Do an 'svn log' at the top of the repository.
	+ Observe that we had created this branch at revision 6.
	+ Now, chdir to ghc66-branch wc and merge all changes from revision 6
	  to HEAD:
	    svn merge -r6:HEAD https://pandoc.googlecode.com/svn/trunk
	+ Make sure that we have no conflicts and commit all changes (as I'm
	  doing right now).

2006-10-27  Recai Oktaş <roktas@debian.org>

	Cleanup cabalize script.

2006-10-27  John MacFarlane <jgm@berkeley.edu>

	Minor wording change in README.

2006-10-27  John MacFarlane <jgm@berkeley.edu>

	Modified package descriptions in debian/control.


src/markdown2pdf.hs
module Main where

import Data.List (isInfixOf, intercalate, isPrefixOf)
import Data.Maybe (isNothing)
import qualified Data.ByteString as BS
import Codec.Binary.UTF8.String (decodeString, encodeString)
import Data.ByteString.UTF8 (toString)
import Control.Monad (unless, guard, liftM, when)
import Control.Concurrent (putMVar, takeMVar, newEmptyMVar, forkIO)
import Control.Exception (tryJust, bracket, evaluate)

import System.IO
import System.IO.Error (isDoesNotExistError)
import System.Environment ( getArgs, getProgName )
import qualified Text.Pandoc.UTF8 as UTF8
import System.Exit (ExitCode (..), exitWith)
import System.FilePath
import System.Directory
import System.Process

-- A variant of 'readProcessWithExitCode' that does not
-- cause an error if the output is not UTF-8. (Copied
-- with slight variants from 'System.Process'.)
readProcessWithExitCode'
    :: FilePath                 -- ^ command to run
    -> [String]                 -- ^ any arguments
    -> String                   -- ^ standard input
    -> IO (ExitCode,String,String) -- ^ exitcode, stdout, stderr
readProcessWithExitCode' cmd args input = do
    (Just inh, Just outh, Just errh, pid) <-
        createProcess (proc cmd args){ std_in  = CreatePipe,
                                       std_out = CreatePipe,
                                       std_err = CreatePipe }

    outMVar <- newEmptyMVar

    -- fork off a thread to start consuming stdout
    out  <- liftM toString $ BS.hGetContents outh
    _ <- forkIO $ evaluate (length out) >> putMVar outMVar ()

    -- fork off a thread to start consuming stderr
    err  <- liftM toString $ BS.hGetContents errh
    _ <- forkIO $ evaluate (length err) >> putMVar outMVar ()

    -- now write and flush any input
    when (not (null input)) $ do hPutStr inh input; hFlush inh
    hClose inh -- done with stdin

    -- wait on the output
    takeMVar outMVar
    takeMVar outMVar
    hClose outh

    -- wait on the process
    ex <- waitForProcess pid

    return (ex, out, err)

run :: FilePath -> [String] -> IO (Either String String)
run file opts = do
  (code, out, err) <- readProcessWithExitCode' (encodeString file)
                        (map encodeString opts) ""
  let msg = out ++ err
  case code of
    ExitFailure _ -> return $ Left  $! msg
    ExitSuccess   -> return $ Right $! msg

parsePandocArgs :: [String] -> IO (Maybe ([String], String))
parsePandocArgs args = do
  result <- run "pandoc" $ ["--dump-args"] ++ args
  return $ either error (parse . map trim . lines) result
  where parse []         = Nothing
        parse ("-":[])   = Just ([], "stdin") -- no output or input
        parse ("-":x:xs) = Just (x:xs, dropExtension x) -- no output
        parse ( x :xs)   = Just (xs,   dropExtension x) -- at least output
        --trim = reverse . dropWhile isSpace . reverse . dropWhile isSpace
        trim = takeWhile (/='\r') . dropWhile (=='\r')

runPandoc :: [String] -> FilePath -> IO (Either String FilePath)
runPandoc inputsAndArgs output = do
  let texFile = addExtension output "tex"
  result <- run "pandoc" $
    ["-s", "--no-wrap", "-r", "markdown", "-w", "latex"]
    ++ inputsAndArgs ++ ["-o", texFile]
  return $ either Left (const $ Right texFile) result

runLatexRaw :: String -> FilePath -> IO (Either (Either String String) FilePath)
runLatexRaw latexProgram file = do
  -- we ignore the ExitCode because pdflatex always fails the first time
  run latexProgram ["-halt-on-error", "-interaction", "nonstopmode",
    "-output-directory", takeDirectory file, dropExtension file] >> return ()
  let pdfFile = replaceExtension file "pdf"
  let logFile = replaceExtension file "log"
  txt <- tryJust (guard . isDoesNotExistError)
         (liftM toString $ BS.readFile logFile)
  let  checks = checkLatex $ either (const "") id txt
  case checks of
  -- err  , bib , ref , msg
    (True , _    , _   , msg) -> return $ Left $ Left msg   -- failure
    (False, True , _   , msg) -> runBibtex file >>
                                (return $ Left $ Right msg) -- citations
    (False, _    , True, msg) -> return $ Left $ Right msg  -- references
    (False, False, False, _ ) -> return $ Right pdfFile     -- success

runLatex :: String -> FilePath -> IO (Either String FilePath)
runLatex latexProgram file = step 3
  where
  step n = do
    result <- runLatexRaw latexProgram file
    case result of
      Left (Left err) -> return $ Left err
      Left (Right _) | n > 1  -> step (n-1 :: Int)
      Right _ | n > 2 -> step (n-1 :: Int)
      Left (Right msg) -> return $ Left msg
      Right pdfFile   -> return $ Right pdfFile

checkLatex :: String -> (Bool, Bool, Bool, String)
checkLatex ""  = (True, False, False, "Could not read log file")
checkLatex txt = (err , bib, ref, unlines $! msgs ++ tips)
  where
  xs `oneOf` x = any (flip isInfixOf x) xs
  msgs = dropWhile (not . errorline) $ lines txt
  errorline ('!':_) = True
  errorline _ = False
  tips = checkPackages msgs
  err = any (oneOf ["!", "LaTeX Error:", "Latex Error:"]) msgs
  bib = any (oneOf ["Warning: Citation"
                   ,"Warning: There were undefined citations"]) msgs
  ref = any (oneOf ["Warning: Reference"
                   ,"Warning: Label"
                   ,"Warning: There were undefined references"
                   ]) msgs

checkPackages :: [String] -> [String]
checkPackages = concatMap chks
  where -- for each message, search 'pks' for matches and give a hint
  chks x = concatMap (chk x) pks
  chk x (k,v) = if sub k `isInfixOf` x then tip k v else []
  sub k   = "`" ++ k ++ ".sty' not found"
  tip k v = ["Please install the '" ++ k ++
             "' package from CTAN:", "  " ++ v]
  pks = [("ucs"
         ,"http://www.ctan.org/tex-archive/macros/latex/contrib/unicode/")
        ,("ulem"
         ,"http://www.ctan.org/tex-archive/macros/latex/contrib/misc/")
        ,("graphicx"
         ,"http://www.ctan.org/tex-archive/macros/latex/required/graphics/")
        ,("fancyhdr"
         ,"http://www.ctan.org/tex-archive/macros/latex/contrib/fancyhdr/")
        ,("array"
         ,"http://www.ctan.org/tex-archive/macros/latex/required/tools/")]

runBibtex :: FilePath -> IO (Either String FilePath)
runBibtex file = do
  let auxFile = replaceExtension file "aux"
  result <- run "bibtex" [auxFile]
  return $ either Left (const $ Right auxFile) result

exit :: String -> IO a
exit x = do
  progName <- getProgName
  UTF8.hPutStrLn stderr $ progName ++ ": " ++ x
  exitWith $ ExitFailure 1

saveStdin :: FilePath -> IO (Either String FilePath)
saveStdin file = do
  text <- liftM toString $ BS.getContents
  UTF8.writeFile file text
  fileExist <- doesFileExist (encodeString file)
  case fileExist of
    False -> return $ Left $! "Could not create " ++ file
    True  -> return $ Right file

saveOutput :: FilePath -> FilePath -> IO ()
saveOutput input output = do
  copyFile (encodeString input) (encodeString output)
  UTF8.hPutStrLn stderr $! "Created " ++ output

main :: IO ()
main = bracket
  -- acquire resource
  (do dir <- getTemporaryDirectory
      let tmp = dir </> "pandoc"
      createDirectoryIfMissing True tmp
      return tmp)

  -- release resource
  ( \tmp -> removeDirectoryRecursive tmp)

  -- run computation
  $ \tmp -> do
    args <- liftM (map decodeString) getArgs
    -- check for invalid arguments and print help message if needed
    let goodopts = ["-f","-r","-N", "-p","-R","-H","-B","-A", "-C","-o","-V"]
    let goodoptslong = ["--from","--read","--strict",
                   "--preserve-tabs","--tab-stop","--parse-raw",
                   "--toc","--table-of-contents", "--xetex", "--luatex",
                   "--number-sections","--include-in-header",
                   "--include-before-body","--include-after-body",
                   "--custom-header","--output",
                   "--template", "--variable",
                   "--csl", "--bibliography", "--data-dir", "--listings"]
    let isOpt ('-':_) = True
        isOpt _       = False
    let opts = filter isOpt args
    -- note that a long option can come in this form: --opt=val
    let isGoodopt x = x `elem` (goodopts ++ goodoptslong) ||
                      any (\o -> (o ++ "=") `isPrefixOf` x) goodoptslong
    let markdown2pdfOpts = ["--xetex","--luatex"]
    unless (all isGoodopt opts) $ do
      (code, out, _err) <- readProcessWithExitCode "pandoc" ["--help"] ""
      UTF8.putStrLn "markdown2pdf [OPTIONS] [FILES]\nOptions:"
      UTF8.putStr $ unlines $
                 filter (\l -> any (`isInfixOf` l) goodoptslong) (lines out)
                 ++ map (replicate 24 ' ' ++) markdown2pdfOpts
      exitWith code

    let args' = filter (`notElem` markdown2pdfOpts) args

    -- check for executable files
    let latexProgram = if "--xetex" `elem` opts
                          then "xelatex"
                          else if "--luatex" `elem` opts
                                  then "lualatex"
                                  else "pdflatex"
    let execs = ["pandoc", latexProgram, "bibtex"]
    paths <- mapM findExecutable execs
    let miss = map snd $ filter (isNothing . fst) $ zip paths execs
    unless (null miss) $ exit $! "Could not find " ++ intercalate ", " miss

    -- parse arguments
    -- if no input given, use 'stdin'
    pandocArgs <- parsePandocArgs args'
    (input, output) <- case pandocArgs of
      Nothing      -> exit "Could not parse arguments"
      Just ([],out) -> do
        stdinFile <- saveStdin (replaceDirectory (takeBaseName out) tmp)
        case stdinFile of
          Left err  -> exit err
          Right f   -> return ([f], out)
      -- no need because we'll pass all arguments to pandoc
      Just (_ ,out) -> return ([], out)
    -- run pandoc
    pandocRes <- runPandoc (input ++ args') $ replaceDirectory output tmp
    case pandocRes of
      Left err -> exit err
      Right texFile  -> do
        -- run pdflatex
        latexRes <- runLatex latexProgram texFile
        case latexRes of
          Left err      -> exit err
          Right pdfFile -> do
            -- save the output creating a backup if necessary
            saveOutput pdfFile $
              replaceDirectory pdfFile (takeDirectory output)

